package trade;

///资金账户
// CThostFtdcTradingAccountField
message AccountInfo {

	///经纪公司代码
	required string	BrokerID = 1;
	///投资者帐号
	required string	AccountID = 2;
	///上次质押金额
	required double	PreMortgage = 3;
	///上次信用额度
	required double	PreCredit = 4;
	///上次存款额
	required double	PreDeposit = 5;
	///上次结算准备金
	required double	PreBalance = 6;
	///上次占用的保证金
	required double	PreMargin = 7;
	///利息基数
	required double	InterestBase = 8;
	///利息收入
	required double	Interest = 9;
	///入金金额
	required double	Deposit = 10;
	///出金金额
	required double	Withdraw = 11;
	///冻结的保证金
	required double	FrozenMargin = 12;
	///冻结的资金
	required double	FrozenCash = 13;
	///冻结的手续费
	required double	FrozenCommission = 14;
	///当前保证金总额
	required double	CurrMargin = 15;
	///资金差额
	required double	CashIn = 16;
	///手续费
	required double	Commission = 17;
	///平仓盈亏
	required double	CloseProfit = 18;
	///持仓盈亏
	required double	PositionProfit = 19;
	///期货结算准备金
	required double	Balance = 20;
	///可用资金
	required double	Available = 21;
	///可取资金
	required double	WithdrawQuota = 22;
	///基本准备金
	required double	Reserve = 23;
	///交易日
	required string	TradingDay = 24;
	///结算编号
	required int32	SettlementID = 25;
	///信用额度
	required double	Credit = 26;
	///质押金额
	required double	Mortgage = 27;
	///交易所保证金
	required double	ExchangeMargin = 28;
	///投资者交割保证金
	required double	DeliveryMargin = 29;
	///交易所交割保证金
	required double	ExchangeDeliveryMargin = 30;

}

///TFtdcPosiDirectionType是一个持仓多空方向类型
enum PosiDirectionType {
	NET = 49;
    LONG = 50;
    SHORT = 51;
}

///TFtdcHedgeFlagType是一个投机套保标志类型
enum HedgeFlagType {
	SPECULATION = 49;
	ARBITRAGE = 50;
	HEDGE = 51;
}

///投资者持仓
// CThostFtdcInvestorPositionField
message PositionInfo {

	///合约代码
	required string	InstrumentID = 1;
	///经纪公司代码
	required string	BrokerID = 2;
	///投资者代码
	required string	InvestorID = 3;
	///持仓多空方向
	required PosiDirectionType	PosiDirection = 4;
	///投机套保标志
	required HedgeFlagType	HedgeFlag = 5;
	///持仓日期
	required string	PositionDate = 6;
	///上日持仓
	required int32	YdPosition = 7;
	///今日持仓
	required int32	Position = 8;
	///多头冻结
	required int32	LongFrozen = 9;
	///空头冻结
	required int32	ShortFrozen = 10;
	///开仓冻结金额
	required double	LongFrozenAmount = 11;
	///开仓冻结金额
	required double	ShortFrozenAmount = 12;
	///开仓量
	required int32	OpenVolume = 13;
	///平仓量
	required int32	CloseVolume = 14;
	///开仓金额
	required double	OpenAmount = 15;
	///平仓金额
	required double	CloseAmount = 16;
	///持仓成本
	required double	PositionCost = 17;
	///上次占用的保证金
	required double	PreMargin = 18;
	///占用的保证金
	required double	UseMargin = 19;
	///冻结的保证金
	required double	FrozenMargin = 20;
	///冻结的资金
	required double	FrozenCash = 21;
	///冻结的手续费
	required double	FrozenCommission = 22;
	///资金差额
	required double	CashIn = 23;
	///手续费
	required double	Commission = 24;
	///平仓盈亏
	required double	CloseProfit = 25;
	///持仓盈亏
	required double	PositionProfit = 26;
	///上次结算价
	required double	PreSettlementPrice = 27;
	///本次结算价
	required double	SettlementPrice = 28;
	///交易日
	required string	TradingDay = 29;
	///结算编号
	required int32	SettlementID = 30;
	///开仓成本
	required double	OpenCost = 31;
	///交易所保证金
	required double	ExchangeMargin = 32;
	///组合成交形成的持仓
	required int32	CombPosition = 33;
	///组合多头冻结
	required int32	CombLongFrozen = 34;
	///组合空头冻结
	required int32	CombShortFrozen = 35;
	///逐日盯市平仓盈亏
	required double	CloseProfitByDate = 36;
	///逐笔对冲平仓盈亏
	required double	CloseProfitByTrade = 37;
	///今日持仓
	required int32	TodayPosition = 38;
	///保证金率
	required double	MarginRateByMoney = 39;
	///保证金率(按手数)
	required double	MarginRateByVolume = 40;

}

///TFtdcOrderPriceTypeType是一个报单价格条件类型
enum OrderPriceTypeType
{
	///任意价
	ANY_PRICE = 49;
	///限价
	LIMIT_PRICE = 50;
	///最优价
	BEST_PRICE = 51;
	///最新价
	LAST_PRICE = 52;
	///最新价浮动上浮1个ticks
	LAST_PRICE_PLUS_ONE_TICKS = 53;
	///最新价浮动上浮2个ticks
	LAST_PRICE_PLUS_TWO_TICKS = 54;
	///最新价浮动上浮3个ticks
	LAST_PRICE_PLUS_THREE_TICKS = 55;
	///卖一价
	ASK_PRICE1 = 56;
	///卖一价浮动上浮1个ticks
	ASK_PRICE1_PLUS_ONE_TICKS = 57;
	///卖一价浮动上浮2个ticks
	ASK_PRICE1_PLUS_TWO_TICKS = 65;
	///卖一价浮动上浮3个ticks
	ASK_PRICE1_PLUS_THREE_TICKS = 66;
	///买一价
	BID_PRICE1 = 67;
	///买一价浮动上浮1个ticks
	BID_PRICE1_PLUS_ONE_TICKS = 68;
	///买一价浮动上浮2个ticks
	BID_PRICE1_PLUS_TWO_TICKS = 69;
	///买一价浮动上浮3个ticks
	BID_PRICE1_PLUS_THREE_TICKS = 70;
}

///TFtdcDirectionType是一个买卖方向类型
enum TradeDirectionType
{
	///买
	BUY = 48;
	///卖
	SELL = 49;
}

///TFtdcTimeConditionType是一个有效期类型类型
enum TimeConditionType
{
	///立即完成，否则撤销
	TC_IOC = 49;
	///本节有效
	TC_GFS = 50;
	///当日有效
	TC_GFD = 51;
	///指定日期前有效
	TC_GTD = 52;
	///撤销前有效
	TC_GTC = 53;
	///集合竞价有效
	TC_GFA = 54;
}

///TFtdcVolumeConditionType是一个成交量类型类型
enum VolumeConditionType
{
	///任何数量
	VC_AV = 49;
	///最小数量
	VC_MV = 50;
	///全部数量
	VC_CV = 51;
}

///TFtdcContingentConditionType是一个触发条件类型
enum ContingentConditionType
{
	///立即
	IMMEDIATELY = 49;
	///止损
	TOUCH = 50;
	///止赢
	TOUCH_PROFIT = 51;
	///预埋单
	PARKED_ORDER = 52;
	///最新价大于条件价
	LAST_PRICE_GREATER_THAN_STOP_PRICE = 53;
	///最新价大于等于条件价
	LAST_PRICE_GREATER_EQUAL_STOP_PRICE = 54;
	///最新价小于条件价
	LAST_PRICE_LESSER_THAN_STOP_PRICE = 55;
	///最新价小于等于条件价
	LAST_PRICE_LESSER_EQUAL_STOP_PRICE = 56;
	///卖一价大于条件价
	ASK_PRICE_GREATER_THAN_STOP_PRICE = 57;
	///卖一价大于等于条件价
	ASK_PRICE_GREATER_EQUAL_STOP_PRICE = 65;
	///卖一价小于条件价
	ASK_PRICE_LESSER_THAN_STOP_PRICE = 66;
	///卖一价小于等于条件价
	ASK_PRICE_LESSER_EQUAL_STOP_PRICE = 67;
	///买一价大于条件价
	BID_PRICE_GREATER_THAN_STOP_PRICE = 68;
	///买一价大于等于条件价
	BID_PRICE_GREATER_EQUAL_STOP_PRICE = 69;
	///买一价小于条件价
	BID_PRICE_LESSER_THAN_STOP_PRICE = 70;
	///买一价小于等于条件价
	BID_PRICE_LESSER_EQUAL_STOP_PRICE = 71;
}

///TFtdcForceCloseReasonType是一个强平原因类型
enum ForceCloseReasonType
{
	///非强平
	NOT_FORCE_CLOSE = 48;
	///资金不足
	LACK_DEPOSIT = 49;
	///客户超仓
	CLIENT_OVER_POSITION_LIMIT = 50;
	///会员超仓
	MEMBER_OVER_POSITION_LIMIT = 51;
	///持仓非整数倍
	NOT_MULTIPLE = 52;
	///违规
	VIOLATION = 53;
	///其它
	FCC_OTHER = 54;
	///自然人临近交割
	PERSON_DELIV = 55;
}

///输入报单
// CThostFtdcInputOrderField
message InputOrder
{
	///经纪公司代码
	required string	BrokerID = 1;
	///投资者代码
	required string	InvestorID = 2;
	///合约代码
	required string	InstrumentID = 3;
	///报单引用
	required string	OrderRef = 4;
	///用户代码
	required string	UserID = 5;
	///报单价格条件
	required OrderPriceTypeType	OrderPriceType = 6;
	///买卖方向
	required TradeDirectionType	Direction = 7;
	///组合开平标志
	required string	CombOffsetFlag = 8;
	///组合投机套保标志
	required string	CombHedgeFlag = 9;
	///价格
	required double	LimitPrice = 10;
	///数量
	required int32	VolumeTotalOriginal = 11;
	///有效期类型
	required TimeConditionType	TimeCondition = 12;
	///GTD日期
	required string	GTDDate = 13;
	///成交量类型
	required VolumeConditionType	VolumeCondition = 14;
	///最小成交量
	required int32	MinVolume = 15;
	///触发条件
	required ContingentConditionType	ContingentCondition = 16;
	///止损价
	required double	StopPrice = 17;
	///强平原因
	required ForceCloseReasonType	ForceCloseReason = 18;
	///自动挂起标志
	required bool	IsAutoSuspend = 19;
	///业务单元
	required string	BusinessUnit = 20;
	///请求编号
	required int32	RequestID = 21;
	///用户强评标志
	required bool	UserForceClose = 22;
};

///TFtdcOrderSubmitStatusType是一个报单提交状态类型
enum OrderSubmitStatusType
{
	///已经提交
	INSERT_SUBMITTED = 48;
	///撤单已经提交
	CANCEL_SUBMITTED = 49;
	///修改已经提交
	MODIFY_SUBMITTED = 50;
	///已经接受
	ACCEPTED = 51;
	///报单已经被拒绝
	INSERT_REJECTED = 52;
	///撤单已经被拒绝
	CANCEL_REJECTED = 53;
	///改单已经被拒绝
	MODIFY_REJECTED = 54;
}

///TFtdcOrderSourceType是一个报单来源类型
enum OrderSourceType
{
	///来自参与者
	PARTICIPANT = 48;
	///来自管理员
	ADMINISTRATOR = 49;
}

///TFtdcOrderStatusType是一个报单状态类型
enum OrderStatusType
{
	///全部成交
	ALL_TRADED = 48;
	///部分成交还在队列中
	PART_TRADED_QUEUEING = 49;
	///部分成交不在队列中
	PART_TRADED_NOT_QUEUEING = 50;
	///未成交还在队列中
	NO_TRADE_QUEUEING = 51;
	///未成交不在队列中
	NO_TRADE_NOT_QUEUEING = 52;
	///撤单
	ORDER_CANCELED = 53;
	///未知
	STATUS_UNKNOWN = 97;
	///尚未触发
	ORDER_NOT_TOUCHED = 98;
	///已触发
	ORDER_TOUCHED = 99;
}

///TFtdcOrderTypeType是一个报单类型类型
enum OrderTypeType
{
	///正常
	NORMAL_ORDER = 48;
	///报价衍生
	DERIVE_FROM_QUOTE = 49;
	///组合衍生
	DERIVE_FROM_COMBINATION = 50;
	///组合报单
	COMBINATION = 51;
	///条件单
	CONDITIONAL_ORDER = 52;
	///互换单
	SWAP_ORDER = 53;
}

///报单
// CThostFtdcOrderField
message Order
{
	///经纪公司代码
	required string	BrokerID = 1;
	///投资者代码
	required string	InvestorID = 2;
	///合约代码
	required string	InstrumentID = 3;
	///报单引用
	required string	OrderRef = 4;
	///用户代码
	optional string	UserID = 5;
	///报单价格条件
	required OrderPriceTypeType	OrderPriceType = 6;
	///买卖方向
	required TradeDirectionType	Direction = 7;
	///组合开平标志
	required string	CombOffsetFlag = 8;
	///组合投机套保标志
	required string	CombHedgeFlag = 9;
	///价格
	required double	LimitPrice = 10;
	///数量
	required int32	VolumeTotalOriginal = 11;
	///有效期类型
	required TimeConditionType	TimeCondition = 12;
	///GTD日期
	optional string	GTDDate = 13;
	///成交量类型
	required VolumeConditionType	VolumeCondition = 14;
	///最小成交量
	required int32	MinVolume = 15;
	///触发条件
	required ContingentConditionType	ContingentCondition = 16;
	///止损价
	optional double	StopPrice = 17;
	///强平原因
	required ForceCloseReasonType	ForceCloseReason = 18;
	///自动挂起标志
	required bool	IsAutoSuspend = 19;
	///业务单元
	optional string	BusinessUnit = 20;
	///请求编号
	optional int32	RequestID = 21;
	// Above fields are same as InputOrder's
	////////////////////////////////////////////////////////////////////

	///本地报单编号
	optional string	OrderLocalID = 22;
	///交易所代码
	optional string	ExchangeID = 23;
	///会员代码
	optional string	ParticipantID = 24;
	///客户代码
	optional string	ClientID = 25;
	///合约在交易所的代码
	optional string	ExchangeInstID = 26;
	///交易所交易员代码
	optional string	TraderID = 27;
	///安装编号
	optional int32	InstallID = 28;
	///报单提交状态
	optional OrderSubmitStatusType	OrderSubmitStatus = 29;
	///报单提示序号
	optional int32	NotifySequence = 30;
	///交易日
	optional string	TradingDay = 31;
	///结算编号
	optional int32	SettlementID = 32;
	///报单编号
	optional string	OrderSysID = 33;
	///报单来源
	optional OrderSourceType	OrderSource = 34;
	///报单状态
	optional OrderStatusType	OrderStatus = 35;
	///报单类型
	optional OrderTypeType	OrderType = 36;
	///今成交数量
	optional int32	VolumeTraded = 37;
	///剩余数量
	optional int32	VolumeTotal = 38;
	///报单日期
	optional string	InsertDate = 39;
	///委托时间
	optional string	InsertTime = 40;
	///激活时间
	optional string	ActiveTime = 41;
	///挂起时间
	optional string	SuspendTime = 42;
	///最后修改时间
	optional string	UpdateTime = 43;
	///撤销时间
	optional string	CancelTime = 44;
	///最后修改交易所交易员代码
	optional string	ActiveTraderID = 45;
	///结算会员编号
	optional string	ClearingPartID = 46;
	///序号
	optional int32	SequenceNo = 47;
	///前置编号
	optional int32	FrontID = 48;
	///会话编号
	optional int32	SessionID = 49;
	///用户端产品信息
	optional string	UserProductInfo = 50;
	///状态信息
	optional string	StatusMsg = 51;
	///用户强评标志
	optional bool	UserForceClose = 52;
	///操作用户代码
	optional string	ActiveUserID = 53;
	///经纪公司报单编号
	optional int32	BrokerOrderSeq = 54;
	///相关报单
	optional string	RelativeOrderSysID = 55;

	/// 附加标识
	optional bool SubmitSuccess = 56;
	optional string SubmitError = 57;
};

///TFtdcTradingRoleType是一个交易角色类型
enum TradingRoleType
{
	///代理
	ER_BROKER = 49;
	///自营
	ER_HOST = 50;
	///做市商
	ER_MAKER = 51;
}

///TFtdcOffsetFlagType是一个开平标志类型
enum OffsetFlagType
{
	///开仓
	OF_OPEN = 48;
	///平仓
	OF_CLOSE = 49;
	///强平
	OF_FORCE_CLOSE = 50;
	///平今
	OF_CLOSE_TODAY = 51;
	///平昨
	OF_CLOSE_YESTERDAY = 52;
	///强减
	OF_FORCE_OFF = 53;
	///本地强平
	OF_LOCAL_FORCE_CLOSE = 54;
}

///TFtdcTradeTypeType是一个成交类型类型
enum TradeTypeType
{
	///普通成交
	TRDT_COMMON = 48;
	///期权执行
	TRDT_OPTIONS_EXECUTION = 49;
	///OTC成交
	TRDT_OTC = 50;
	///期转现衍生成交
	TRDT_EFP_DERIVED = 51;
	///组合衍生成交
	TRDT_COMBINATION_DERIVED = 52;
}

///TFtdcPriceSourceType是一个成交价来源类型
enum PriceSourceType
{
	///前成交价
	PSRC_LAST_PRICE = 48;
	///买委托价
	PSRC_BUY = 49;
	///卖委托价
	PSRC_SELL = 50;
}

///成交
// CThostFtdcTradeField
message Trade
{
	///经纪公司代码
	required string	BrokerID = 1;
	///投资者代码
	required string	InvestorID = 2;
	///合约代码
	required string	InstrumentID = 3;
	///报单引用
	required string	OrderRef = 4;
	///用户代码
	required string	UserID = 5;
	///交易所代码
	required string	ExchangeID = 6;
	///成交编号
	required string	TradeID = 7;
	///买卖方向
	required TradeDirectionType	Direction = 8;
	///报单编号
	required string	OrderSysID = 9;
	///会员代码
	required string	ParticipantID = 10;
	///客户代码
	required string	ClientID = 11;
	///交易角色
	required TradingRoleType	TradingRole = 12;;
	///合约在交易所的代码
	required string	ExchangeInstID = 13;
	///开平标志
	required OffsetFlagType	OffsetFlag = 14;
	///投机套保标志
	required HedgeFlagType	HedgeFlag = 15;
	///价格
	required double	Price =  16;
	///数量
	required int32	Volume = 17;
	///成交时期
	required string	TradeDate = 18;
	///成交时间
	required string	TradeTime = 19;
	///成交类型
	required TradeTypeType	TradeType = 20;
	///成交价来源
	required PriceSourceType	PriceSource = 21;
	///交易所交易员代码
	required string	TraderID = 22;
	///本地报单编号
	required string	OrderLocalID = 23;
	///结算会员编号
	required string	ClearingPartID = 24;
	///业务单元
	required string	BusinessUnit = 25;
	///序号
	required int32	SequenceNo = 26;
	///交易日
	required string	TradingDay = 27;
	///结算编号
	required int32	SettlementID = 28;
	///经纪公司报单编号
	required int32	BrokerOrderSeq = 29;
};


///TFtdcActionFlagType是一个操作标志类型
enum ActionFlagType
{
	///删除
	AF_Delete = 48;
	///修改
	AF_Modify = 51;
};


///输入报单操作
// CThostFtdcInputOrderActionField
message InputOrderAction
{
	///经纪公司代码
	required string	BrokerID = 1;
	///投资者代码
	required string	InvestorID = 2;
	///报单操作引用
	required int32	OrderActionRef = 3;
	///报单引用
	required string	OrderRef = 4;
	///请求编号
	required int32	RequestID = 5;
	///前置编号
	required int32	FrontID = 6;
	///会话编号
	required int32	SessionID = 7;
	///交易所代码
	required string	ExchangeID = 8;
	///报单编号
	required string	OrderSysID = 9;
	///操作标志
	required ActionFlagType	ActionFlag = 10;
	///价格
	required double	LimitPrice = 11;
	///数量变化
	required int32 VolumeChange = 12;
	///用户代码
	required string	UserID = 13;
	///合约代码
	required string	InstrumentID = 14;
}

message MultiLegOrder
{
	required string OrderId = 1;
	required string PortfolioId = 2;
	required int32 Quantity = 3;
	
	repeated Order Legs = 4;
	required string OpenOrderId = 5;
	required string OpenDate = 6;
}
